FirebaseMLModelDownloader/Tests/Unit/ModelDownloaderUnitTests.swift

```diff
@@ -298,61 +291,87 @@ class NetworkingUnitTests: XCTestCase {
                                                    fileURL: tempFileURL)))
-
     wait(for: [completionExpectation], timeout: 0.5)
 
-    // Cleanup temp file.
     try? ModelFileManager.removeFile(at: tempFileURL)
   }
 
-  func testModelDownloadWithSuccessfulRetry() {
-    let session = fakeModelInfoSessionWithURL(fakeDownloadURL, statusCode: 200)
-    let modelInfoRetriever = fakeModelRetriever(fakeSession: session)
+  func testModelDownloadWith400() {
     let tempFileURL = tempFile()
     let remoteModelInfo = fakeModelInfo()
-
-    var fakeResponse = HTTPURLResponse(url: fakeFileURL,
+    let fakeResponse = HTTPURLResponse(url: fakeFileURL,
                                        statusCode: 400,
                                        httpVersion: nil,
                                        headerFields: nil)!
-
-    // Expect file downloader to be called.
     let downloader = MockModelFileDownloader()
-    let fileDownloaderExpectation = expectation(description: "file downloader")
-
+    let fileDownloaderExpectation = expectation(description: "File downloader")
     downloader.downloadFileHandler = { url in
       fileDownloaderExpectation.fulfill()
       XCTAssertEqual(url, self.fakeDownloadURL)
-      /// Update model info response to be successful this time.
-      fakeResponse = HTTPURLResponse(url: self.fakeFileURL,
-                                     statusCode: 200,
-                                     httpVersion: nil,
-                                     headerFields: nil)!
     }
 
-    let completionExpectation = expectation(description: "completion handler")
-
+    let completionExpectation = expectation(description: "Completion handler")
     let modelDownloadTask = ModelDownloadTask(remoteModelInfo: remoteModelInfo,
                                               appName: "fakeAppName",
                                               defaults: .createTestInstance(testName: #function),
-                                              downloader: downloader,
-                                              modelInfoRetriever: modelInfoRetriever) { result in
+                                              downloader: downloader)
+    modelDownloadTask.download(progressHandler: nil) { result in
+      completionExpectation.fulfill()
+      switch result {
+      case .success: XCTFail("Unexpected successful model download.")
+      case let .failure(error):
+        XCTAssertEqual(error, DownloadError.failedPrecondition)
+      }
+    }
+    wait(for: [fileDownloaderExpectation], timeout: 0.1)
+
+    downloader
+      .completion?(.success(FileDownloaderResponse(urlResponse: fakeResponse,
+                                                   fileURL: tempFileURL)))
+    wait(for: [completionExpectation], timeout: 0.5)
+
+    try? ModelFileManager.removeFile(at: tempFileURL)
+  }
+
+  func testGetModelSuccessfulRetry() {
+    let modelDownloader = ModelDownloader.modelDownloader()
```

Comment:
nit: Sometimes I find it useful to annotate steps of a long test case with comments like here:  https://github.com/firebase/firebase-ios-sdk/blob/bab9af811f165a53f15402dc7526830c265f4155/FirebaseInstallations/Source/Tests/Unit/FIRInstallationsIDControllerTests.m#L156 It may be helpful here as well.

Reply:
Nice! I've added some basic comments for now, but I'll try to be this elaborate in a future PR.